### ISSUE TITLE
[Feature Request] Open video files in local player

### ISSUE THREAD

**Issue by tissole** [kotatogram/kotatogram-desktop#207]
### Is your feature request related to a problem?
Currently, the video player can render only .mp4 and .mov files. MKV, AVI, FLV, TS can't be played on the video player.

### Describe the solution you'd like
Add an option for the users to open video files in the local video player. Not only a greater variety of files could be played directly, but the process is more smoothly and battery friendly. Now the video player consumes more resources than the local video player.

**Comment by ilya-fedin** [comment]
Such option already exists, look at kotatogram settings

**Comment by tissole** [comment]
??? I searched all settings, I could not find this option in any menu. Where exactly is this option located?

**Comment by ilya-fedin** [comment]
Ah, it's not released yet. You still can enable it via `videoplayer` cheat code inherited from tdesktop.

**Comment by tissole** [comment]
Please give more details, I don't know anything about this cheat code.
P.S. When this feature is coming?

**Comment by ilya-fedin** [comment]
> Please give more details, I don't know anything about this cheat code.

Go to the main settings view (where `Notifications`, `Chat Settings`, etc), enter `videoplayer` blindly (like a cheat code in a game)

> P.S. When this feature is coming?

It's already present on dev branch. No ETA for next pointing release.

**Comment by ilya-fedin** [comment]
Here's the list of tdesktop cheat codes: https://github.com/telegramdesktop/tdesktop/wiki/Hidden-modes

**Comment by tissole** [comment]
Aaaa...the treasure trove. Thanks a lot!

**Comment by tissole** [comment]
I tested but unfortunately, it does not work, It downloads the file instead to open it in the external player. I will wait for the new version.

Is it also possible to open Youtube links in the external player? Now, these links are opened in the browser, which is quite annoying.

**Comment by ilya-fedin** [comment]
> It downloads the file instead to open it in the external player.

Of course it downloads, you can't open something in external program without downloading.

> I will wait for the new version.

It just adds a visible switch for that feature. Nothing else.

**Comment by tissole** [comment]
Oh! I was expecting the file to be streamed, not downloaded. That's disappointing...
If that's not going to work, what about making the internal player stream, mkv, avi files? Or is a Telegram limitation that can't be bypassed?

The desktop client can be made to play directly Youtube links as the mobile client does?

**Comment by ilya-fedin** [comment]
I don't imagine how to stream to external players. Usually system provides you an API like ShellExecute/xdg-open and all you can is to specify local file, then it finds the appropriate application that can handle that file and opens it. They don't provide any API to handle streams.

**Comment by ilya-fedin** [comment]
> The desktop client can be made to play directly Youtube links as the mobile client does?

That would require a lot of effort, personally I'm not interested in this, so I won't work on this

**Comment by ilya-fedin** [comment]
> If that's not going to work, what about making the internal player stream, mkv, avi files? Or is a Telegram limitation that can't be bypassed?

I'm not sure we should do that. I think this should be discussed with @RadRussianRus

**Comment by RadRussianRus** [comment]
I don't think it's going to work. Even if player supports streaming, it's probably by HTTP(S), and Telegram uses MTProto.

It could be solved with embedding of HTTP server, so video can be converted from MTProto to HTTP, but that's totally overkill.

Another way is to make a player plugin with embedded TDLib to support streaming. But I'm not sure if TDLib even supports streaming. Though even if it does, you'll need to make a plugin for every player you need, and maintain it (at least in terms of updating TDLib). And with Telegram's pretty unstable APIs it probably will be painful.

So it's either overkill or painful to make and maintain (also first way could be both overkill and painful).

**Comment by tissole** [comment]
Thanks for the answers! If making Youtube links play in the external player is difficult, what about this:
Opening any links with a custom URI Scheme like in this fork
https://github.com/Forkgram/tdesktop

> It could be solved with embedding of HTTP server, so video can be converted from MTProto to HTTP, but that's totally overkill.

I see. That's OK, I found a solution based on this concept, streaming Telegram files through an HTTP server.

**Comment by ilya-fedin** [comment]
> Opening any links with a custom URI Scheme like in this fork

WTF

**Comment by tissole** [comment]
![hEhGMj3MPS](https://user-images.githubusercontent.com/78413495/128864680-94e48e99-bf21-470e-bb55-30079996ebb1.png)

![R9FWHCpsUg](https://user-images.githubusercontent.com/78413495/128864709-8cfd5ef5-8f09-419e-9786-6806a0eb5ccc.png)

**Comment by ilya-fedin** [comment]
What does that do?

**Comment by tissole** [comment]
![Forkgram_hErKGJDtLN](https://user-images.githubusercontent.com/78413495/128873174-6aa78c4a-b1d4-4140-b03e-8a9b384f99f3.png)

**Comment by ilya-fedin** [comment]
It doesn't make clear what happens when you click "Open Link with URI scheme"

**Comment by tissole** [comment]
It opens the player (Potplayer), but it does not play the video for some reason. Or is something wrong with the implementation, or I did not set this correctly.

**Comment by ilya-fedin** [comment]
How does it understand what app to open?

**Comment by tissole** [comment]
Supposedly through this URI scheme `potplayer://`

**Comment by ilya-fedin** [comment]
so you specify URI scheme in settings and it just opens that URI scheme via system handler by that menu item?

**Comment by tissole** [comment]
Yes, but only for Potplayer, I tried `vlc://`, and is not working. So I don't know what the scheme is for specific players. I will open a ticket to ask the developers.

**Comment by ilya-fedin** [comment]
> So I don't know what the scheme is for specific players.

Players usually don't provide schemes. So I think such a feature is useless.

**Comment by tissole** [comment]
They updated the fork and the feature is working now. I tried with ` potplayer://` and` mpv://`. I also found a workaround, drop the link in the player :)
